### PR TITLE
[PyTorch] Stack-allocate boxed args for RecordFunction

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76266

Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

### CODE REVIEW

**Comment by swolchok** [comment]
An interesting follow-up would be to make it possible to avoid heap-allocating a Stack for calls to boxed KernelFunctions made via make_boxed_from_unboxed_functor. I have no idea how common this is, though.

File: aten/src/ATen/core/dispatch/Dispatcher.h [hunk 1]
```diff
@@ -490,6 +490,52 @@ struct CaptureKernelCall<void> {
...
+static inline constexpr size_t boxed_size() {
+  return BoxedSize<Args...>::value;
+}
+
+using IValueStorage = std::aligned_storage_t<sizeof(IValue), alignof(IValue)>;
+
+template <typename T>
+C10_DISPATCHER_INLINE_UNLESS_MOBILE void box(IValueStorage* dest, T& arg, int& lastIdx) {
+  new (&dest[lastIdx]) IValue(arg);
+  lastIdx++;
```
Comment (ezyang) [inline aten/src/ATen/core/dispatch/Dispatcher.h:521]: This needs to be kept consistent with our other boxing logic, is that right?

Comment (ezyang) [reply aten/src/ATen/core/dispatch/Dispatcher.h:521]: I'd suggest a cross-ref with aten/src/ATen/core/boxing/impl/boxing.h (or perhaps moving it to this file entirely)

File: aten/src/ATen/core/dispatch/Dispatcher.h [hunk 2]
```diff
@@ -490,6 +490,52 @@ struct CaptureKernelCall<void> {
...
+
+template <class... Args>
+static inline constexpr size_t boxed_size() {
+  return BoxedSize<Args...>::value;
+}
+
+using IValueStorage = std::aligned_storage_t<sizeof(IValue), alignof(IValue)>;
+
+template <typename T>
+C10_DISPATCHER_INLINE_UNLESS_MOBILE void box(IValueStorage* dest, T& arg, int& lastIdx) {
```
Comment (ezyang) [inline aten/src/ATen/core/dispatch/Dispatcher.h:519]: The name here is a bit ambiguous; what we're doing is placement new boxing; would be nice if the name had this so people don't try to use this invalidly.

**Review by ezyang — APPROVED** [review]
mechanically looks all fine, just some naming / file placement stuff

File: aten/src/ATen/core/dispatch/Dispatcher.h (hunk 2 above)
Comment (swolchok) [reply aten/src/ATen/core/dispatch/Dispatcher.h:519]: it's in detail and takes a weird typedef; they'll have trouble using it without writing obvious "here be dragons" stuff like aligned_storage

**Pull request closed without merging**